### ISSUE TITLE
feat: raw_entry_mut: iterate all entries raw?

### ISSUE THREAD

**Issue by CAD97** [rust-lang/hashbrown#161]
TL;DR: add functionality roughly isomorphic to

```rust
fn RawEntryBuilderMut::all(self) -> impl Iterator<Item=RawOccupiedEntryMut>;
```

---

I currently have a cache with roughly the shape

```rust
struct Cache {
    hasher: ahash::RandomState, // dedupe the 2×u64 hasher state and enforce custom hashing
    nodes: HashMap<Arc<Node>, (), ()>,
    tokens: HashMap<Arc<Token>, (), ()>,
}
```

which makes heavy use of the `raw_entry_mut` API to get allocation-free cache hits even though `Node` and `Token` are `?Sized` types, and to do custom hashing for items in the cache (for example, only checking "thin" equivalence for nodes rather than recursing into node children).

I would like to add a garbage collection option to the cache to drop any arc in the cache that is not used outside of the cache. If these were "normal" hashmaps/hashsets, this would be trivial:

```rust
pub fn Cache::gc(&mut self) {
    self.nodes.retain(|(node, ())| Arc::strong_count(node) > 1);
    self.tokens.retain(|(token, ())| Arc::strong_count(token) > 1);
}
```

but because `retain` requires `S: BuildHasher` and my `raw_entry` shenanigans, this doesn't work on its own.

So: is it possible to remove the `S: BuildHasher` requirement from `HashMap::retain` or expose a `raw_entry_mut` API to iterate over all entries in `raw_entry` mode?

```rust
pub fn Cache::gc(&mut self) {
    for entry in self.nodes.raw_entry_mut().all() {
        if Arc::strong_count(entry.key()) == 1 {
            entry.remove_entry();
        }
    }
    for entry in self.tokens.raw_entry_mut().all() {
        if Arc::strong_count(entry.key()) == 1 {
            entry.remove_entry();
        }
    }
}
```

**Comment by CAD97** [comment]
A local test shows that for both `retain` and `drain_filter` currently require `K: Eq + Hash, S: BuildHasher`, just moving them into the `impl` block without any bounds on `<K, V, S>` continues to work and pass all tests without any further adjustments.

**Comment by Amanieu** [comment]
That sounds like a good idea, would you like to send a PR?

**Comment by CAD97** [comment]
Based on some quick experimentation, because `RawOccupiedEntryMut` holds a `&mut RawTable`, any `.raw_entry_mut().all()` would have to be a streaming iterator rather than a standard iterator.
